File: content/snippets/go/language_basics/recursion.go
package main

import "fmt"

func factorial(n int) int {
	if n <= 1 {
		return 1
	}
	return n * factorial(n-1)
}

func sumRecursive(values []int, index int) int {
	if index >= len(values) {
		return 0
	}
	return values[index] + sumRecursive(values, index+1)
}

func main() {
	values := []int{1, 2, 3, 4}
	fmt.Println("factorial(5):", factorial(5))
	fmt.Println("sumRecursive(values):", sumRecursive(values, 0))
}

File: content/snippets/cpp/language_basics/recursion.cpp
#include <iostream>
#include <vector>

int factorial(int n) {
    if (n <= 1) {
        return 1;
    }
    return n * factorial(n - 1);
}

int sum_recursive(const std::vector<int>& values, size_t index = 0) {
    if (index >= values.size()) {
        return 0;
    }
    return values[index] + sum_recursive(values, index + 1);
}

int main() {
    std::vector<int> values{1, 2, 3, 4};
    std::cout << "factorial(5): " << factorial(5) << "\n";
    std::cout << "sum_recursive(values): " << sum_recursive(values) << "\n";
    return 0;
}
